File: ruleMatch/src/main/scala/com/bigdata/rulematch/scala/function/RuleMatchKeyedProcessFunctionV3.scala
package com.bigdata.rulematch.scala.function

import java.sql.Connection

import com.bigdata.rulematch.scala.bean.rule.{EventCondition, EventSeqCondition}
import com.bigdata.rulematch.scala.bean.{EventLogBean, RuleMatchResult}
import com.bigdata.rulematch.scala.conf.EventRuleConstant
import com.bigdata.rulematch.scala.datagen.RuleConditionEmulator
import com.bigdata.rulematch.scala.router.SimpleQueryRouter
import com.bigdata.rulematch.scala.service.impl.{ClickHouseQueryServiceImpl, HBaseQueryServiceImpl}
import com.bigdata.rulematch.scala.utils.{ConnectionUtils, EventRuleCompareUtils}
import org.apache.commons.configuration2.PropertiesConfiguration
import org.apache.commons.dbutils.DbUtils
import org.apache.flink.configuration.Configuration
import org.apache.flink.streaming.api.functions.KeyedProcessFunction
import org.apache.flink.util.Collector
import org.slf4j.{Logger, LoggerFactory}

/**
 *
 * 静态规则匹配KeyedProcessFunction 版本3
 *
 * 主要是把查询的流程放到一个路由器类中进行管理
 *
 * @author Administrator
 * @version 1.0
 * @date 2021-12-19 15:07
 */
class RuleMatchKeyedProcessFunctionV3 extends KeyedProcessFunction[String, EventLogBean, RuleMatchResult] {
  private val logger: Logger = LoggerFactory.getLogger(this.getClass.getName.stripSuffix("$"))

  private val config: PropertiesConfiguration = EventRuleConstant.config

  private var queryRouter: SimpleQueryRouter = null

  override def open(parameters: Configuration): Unit = {

    // 初始化查询路由对象
    queryRouter = new SimpleQueryRouter()
  }

  override def processElement(event: EventLogBean,
                              ctx: KeyedProcessFunction[String, EventLogBean, RuleMatchResult]#Context,
                              out: Collector[RuleMatchResult]): Unit = {

    //获取模拟生成的规则
    val ruleCondition = RuleConditionEmulator.getRuleConditions()

    logger.debug(s"获取到的规则条件: ${ruleCondition}")

    //判断是否满足规则触发条件
    val isMatch = queryRouter.ruleMatch(event, ctx.getCurrentKey, ruleCondition)

    if (isMatch) {

      logger.info("所有规则匹配成功,准备输出匹配结果信息...")

      //创建规则匹配结果对象
      val matchResult = RuleMatchResult(ctx.getCurrentKey, ruleCondition.ruleId, event.timeStamp, System.currentTimeMillis())

      //将匹配结果输出
      out.collect(matchResult)
    }

  }

  override def close(): Unit = {
    //关闭连接
    queryRouter.closeConnection()
  }
}

File: ruleMatch/src/main/java/com/bigdata/rulematch/java/function/RuleMatchKeyedProcessFunctionV3.java
package com.bigdata.rulematch.java.function;

import com.bigdata.rulematch.java.bean.EventLogBean;
import com.bigdata.rulematch.java.bean.RuleMatchResult;
import com.bigdata.rulematch.java.bean.rule.RuleCondition;
import com.bigdata.rulematch.java.conf.EventRuleConstant;
import com.bigdata.rulematch.java.datagen.RuleConditionEmulator;
import com.bigdata.rulematch.java.router.RuleMatchRouter;
import org.apache.commons.configuration2.PropertiesConfiguration;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.streaming.api.functions.KeyedProcessFunction;
import org.apache.flink.util.Collector;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.sql.SQLException;
import java.text.ParseException;


/**
 * 封装规则匹配功能路由,便于后续控制匹配流程
 *
 * @author HanDongfang
 * @version 1.0
 * @create 2021-12-27  14:47
 */
public class RuleMatchKeyedProcessFunctionV3 extends KeyedProcessFunction<String, EventLogBean, RuleMatchResult> {
    private Logger logger = LoggerFactory.getLogger(this.getClass().getName());

    private PropertiesConfiguration config = EventRuleConstant.config;
    /**
     * 规则查询路由
     */
    private RuleMatchRouter ruleMatchRouter = null;


    @Override
    public void open(Configuration parameters) throws SQLException, IOException, ClassNotFoundException {
        // 初始化规则查询路由
        ruleMatchRouter = new RuleMatchRouter();
    }


    @Override
    public void processElement(EventLogBean eventLogBean, Context context, Collector<RuleMatchResult> collector) {
        //1. 获取模拟生成的规则
        RuleCondition ruleCondition = null;
        try {
            ruleCondition = RuleConditionEmulator.getRuleConditions(eventLogBean.getTimeStamp());
            logger.debug(String.format("获取到的规则条件: %s", ruleCondition));
        } catch (ParseException e) {
            logger.error("规则生成出错:", e);
        }

        if (ruleCondition != null) {
            //keyBy的值要获取到
            String keyByFiedValue = context.getCurrentKey();

            boolean isMatch = ruleMatchRouter.ruleMatch(eventLogBean, keyByFiedValue, ruleCondition);

            if (isMatch) {
                logger.debug("所有规则匹配成功,准备输出匹配结果信息...");

                //创建规则匹配结果对象
                RuleMatchResult matchResult = new RuleMatchResult("rule-001", "规则1", eventLogBean.getTimeStamp(), System.currentTimeMillis());

                //将匹配结果输出
                collector.collect(matchResult);
            }
        }
    }

    @Override
    public void close() {
        //关闭query连接
        ruleMatchRouter.closeConnection();
    }

}
